### ISSUE TITLE
Crash when maximizing buffer

### ISSUE THREAD

**Issue by fdev31** [neovide/neovide#1669]
**Describe the bug**
neovide seems to crash when maximizing the current buffer

**To Reproduce**
1. Open Neovide
2. :split
3. Ctrl+W, _
4. *Crash*

**Expected behavior**
Selected buffer is maximized

**Screenshots**
Not applicable

**Desktop (please complete the following information):**
 - OS: Archlinux @ 9/12/2022
 - Neovide Version 0.10.3
 - Neovim Version 0.8.1
 
**Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**

```
DEBUG [neovide::bridge::command] Starting neovim with: Command { std: "/usr/bin/nvim" "--embed" "-p", kill_on_drop: false }
INFO [neovide::bridge::setup] Neovide registered to nvim with channel id 1
INFO [neovide::bridge] Neovim process attached
DEBUG [neovide::settings::window_geometry] Loaded window settings: Windowed { position: PhysicalPosition { x: 2, y: 530 }, size: Dimensions { width: 95, height: 24 } }
DEBUG [neovide::bridge::events] Ignored style attribute: "nocombine"
DEBUG [neovide::bridge::events] Ignored style attribute: "nocombine"
DEBUG [neovide::bridge::events] Ignored style attribute: "nocombine"
DEBUG [neovide::bridge::events] Ignored style attribute: "nocombine"
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_size: 18.67px
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_width: 11.49px
DEBUG [neovide::renderer::fonts::caching_shaper] Font width: 11.49px (avg: 11.49px)
DEBUG [neovide::renderer::fonts::caching_shaper] Fudge factor: 0.96
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font size: 17.88px
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font width: 11.00px
INFO [neovide::window] window created (scale_factor: 1.0000, font_dimensions: Dimensions { width: 11, height: 22 })
DEBUG [neovide::renderer::fonts::caching_shaper] Updating font: 
DEBUG [neovide::renderer::fonts::caching_shaper] Font updated to: 
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_size: 18.67px
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_width: 11.49px
DEBUG [neovide::renderer::fonts::caching_shaper] Font width: 11.49px (avg: 11.49px)
DEBUG [neovide::renderer::fonts::caching_shaper] Fudge factor: 0.96
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font size: 17.88px
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font width: 11.00px
DEBUG [neovide::renderer::fonts::caching_shaper] Updating font: Fira Code,Noto Color Emoji:h12:#e-subpixelantialias
DEBUG [neovide::renderer::fonts::caching_shaper] Font updated to: Fira Code,Noto Color Emoji:h12:#e-subpixelantialias
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_size: 16.00px
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_width: 9.85px
DEBUG [neovide::renderer::fonts::caching_shaper] Font width: 9.85px (avg: 9.85px)
DEBUG [neovide::renderer::fonts::caching_shaper] Fudge factor: 1.02
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font size: 16.25px
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font width: 10.00px
INFO [neovide::running_tracker] Quit with code 0: Quit from neovim
DEBUG [neovide::settings::window_geometry] Saved Window Settings: {"window":{"Windowed":{"position":{"x":2,"y":530},"size":{"width":95,"height":24}}}}
```

**Comment by MultisampledNight** [comment]
Thanks for reporting! Unfortunately though I'm unable to reproduce this, neither with my config nor with no config.

Could you try running Neovide through `neovide -- -u NORC` to see whether you can reproduce this without config? The logs mention that the underlying Neovim instance closed and in turn Neovide itself, maybe a mapping/plugin is behaving in an unexpected way.

**Comment by fdev31** [comment]
It seems the line required to produce the bug isn't plugin related:
I can run with
```
neovide --multigrid -- -u NORC
```
then set wmh:
```
:set wmh=0
```
and follow the bug scenario to have a crash.

**Comment by MultisampledNight** [comment]
...okay, that's really interesting. If it doesn't obstruct your workflow too much, you can work around the fullscreen behavior with `:only`.

**Comment by MultisampledNight** [comment]
Nope, Neovim doesn't exit. It's really Neovide crashing, this is the backtrace for anyone interested (produced by running `neovide --nofork` on a debug build):

<details>
```
thread 'main' panicked at 'Could not create surface', src/renderer/rendered_window.rs:83:6
stack backtrace:
   0: rust_begin_unwind
             at /rustc/23415176968e81e0aac92d0218612a89c4e68a82/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/23415176968e81e0aac92d0218612a89c4e68a82/library/core/src/panicking.rs:64:14
   2: core::panicking::panic_display
             at /rustc/23415176968e81e0aac92d0218612a89c4e68a82/library/core/src/panicking.rs:135:5
   3: core::panicking::panic_str
             at /rustc/23415176968e81e0aac92d0218612a89c4e68a82/library/core/src/panicking.rs:119:5
   4: core::option::expect_failed
             at /rustc/23415176968e81e0aac92d0218612a89c4e68a82/library/core/src/option.rs:1879:5
   5: core::option::Option<T>::expect
             at /rustc/23415176968e81e0aac92d0218612a89c4e68a82/library/core/src/option.rs:741:21
   6: neovide::renderer::rendered_window::build_window_surface
             at ./src/renderer/rendered_window.rs:74:5
   7: neovide::renderer::rendered_window::build_window_surface_with_grid_size
             at ./src/renderer/rendered_window.rs:91:23
   8: neovide::renderer::rendered_window::RenderedWindow::handle_window_draw_command
             at ./src/renderer/rendered_window.rs:400:43
   9: neovide::renderer::Renderer::handle_draw_command
             at ./src/renderer/mod.rs:248:25
  10: neovide::renderer::Renderer::draw_frame
             at ./src/renderer/mod.rs:153:13
  11: neovide::window::GlutinWindowWrapper::draw_frame
             at ./src/window/mod.rs:223:17
  12: neovide::window::create_window::{{closure}}
             at ./src/window/mod.rs:514:13
  13: winit::platform_impl::platform::sticky_exit_callback
             at /home/multisn8/.cargo/git/checkouts/winit-58efae9df3f5264a/27967af/src/platform_impl/linux/mod.rs:760:5
  14: winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run_return
             at /home/multisn8/.cargo/git/checkouts/winit-58efae9df3f5264a/27967af/src/platform_impl/linux/wayland/event_loop/mod.rs:388:13
  15: winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run
             at /home/multisn8/.cargo/git/checkouts/winit-58efae9df3f5264a/27967af/src/platform_impl/linux/wayland/event_loop/mod.rs:191:9
  16: winit::platform_impl::platform::EventLoop<T>::run
             at /home/multisn8/.cargo/git/checkouts/winit-58efae9df3f5264a/27967af/src/platform_impl/linux/mod.rs:676:56
  17: winit::event_loop::EventLoop<T>::run
             at /home/multisn8/.cargo/git/checkouts/winit-58efae9df3f5264a/27967af/src/event_loop.rs:154:9
  18: neovide::window::create_window
             at ./src/window/mod.rs:470:5
  19: neovide::main
             at ./src/main.rs:145:5
  20: core::ops::function::FnOnce::call_once
             at /rustc/23415176968e81e0aac92d0218612a89c4e68a82/library/core/src/ops/function.rs:507:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
</details>

The `could not create surface` thing seems somehow known to me in other issues, a short way to reproduce this is awesome. Thank you!

**Comment by MultisampledNight** [comment]
Found it. Surfaces with a height/width of 0 summon dragons (paraphrasing "causing the crash"), specifically here:
https://github.com/neovide/neovide/blob/225a5f5df5d1fc9a3f914bd0f579c673cd5d6c74/src/renderer/rendered_window.rs#L400-L404

The `build_window_surface` function called down in the stack:

https://github.com/neovide/neovide/blob/225a5f5df5d1fc9a3f914bd0f579c673cd5d6c74/src/renderer/rendered_window.rs#L61

should probably not even take a `i32` anyways, but rather a `NonZero<u32>` or the like. I'll draft up a PR.

**Comment by MultisampledNight** [comment]
Maybe this is a skia bug, the [docs to `Surface::new_render_target`](https://api.skia.org/classSkSurface.html#aa32278bea58b9f229190c45b59f01548) mention:

> [...]; width, or height, or both, may be zero

But on the other hand that's really unlikely. Zero-sized surfaces can happen easily, why are they such a problem exactly here?

**Comment by MultisampledNight** [comment]
I feel stupid. `None` refers to the internal skia creation function returning a null pointer, which is likely valid for a zero-sized surface. So I guess we need to instead accurately hold a not-yet-existent surface instead.

**Comment by MultisampledNight** [comment]
...but the docs mention

> Returns [SkSurface](https://api.skia.org/classSkSurface.html) if all parameters are valid; otherwise, nullptr

a

**Comment by fdev31** [comment]
forcing non-zero values in pixel_size fixes/works-around this issue. I can't notice any regression with this change.

**Comment by fdev31** [comment]
There is a workaround for months:
```rust
fn build_window_surface_with_grid_size(
    parent_canvas: &mut Canvas,
    grid_renderer: &GridRenderer,
    mut grid_size: Dimensions,
) -> Surface {
    if grid_size.height == 0 {
        grid_size.height = 1;
    }
```

(in rendered_window.rs)

Can we apply it, at least temporarily until the library is fixed ?

**Comment by fredizzimo** [comment]
This should be fixed in main, by https://github.com/neovide/neovide/pull/1789

**Comment by fdev31** [comment]
I can reproduce it using neovide git version (0.11.2 is unaffected).

**Comment by fredizzimo** [comment]
The original bug was fixed, but now it crashes in a different place with this backtrace
```
   0: neovide::main::{{closure}}
             at src/main.rs:74:25
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/alloc/src/boxed.rs:2021:9
      std::panicking::rust_panic_with_hook
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:711:13
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:597:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:170:18
   4: rust_begin_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:595:5
   5: core::panicking::panic_fmt
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
   6: core::panicking::panic
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:117:5
   7: core::num::<impl isize>::rem_euclid
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/num/int_macros.rs:2137:21
   8: neovide::utils::ring_buffer::RingBuffer<T>::get_array_index
             at src/utils/ring_buffer.rs:87:9
   9: neovide::utils::ring_buffer::RingBuffer<T>::resize
             at src/utils/ring_buffer.rs:75:21
  10: neovide::editor::grid::CharacterGrid::resize
             at src/editor/grid.rs:45:9
  11: neovide::editor::window::Window::resize
             at src/editor/window.rs:115:9
  12: neovide::editor::Editor::resize_window
             at src/editor/mod.rs:296:13
  13: neovide::editor::Editor::handle_editor_command
             at src/editor/mod.rs:167:21
  14: neovide::editor::start_editor::{{closure}}
             at src/editor/mod.rs:553:13
```

**Comment by fredizzimo** [comment]
This was fixed by #2150
